Update OAuth2 OIDC SDK

AthenaEryma opened the pull request:
This commit updates the Nimbus OAuth2 OIDC SDK and the associated Nimbus JOSE+JWT, however a few odd choices had to be made in the process.

First, we update to versions which are old at time of merge. This is because versions of Nimbus JOSE+JWT 9.38 and after through time of writing [contain a bug](https://bitbucket.org/connect2id/nimbus-jose-jwt/issues/550/java-module-doesnt-work-properly-with) in which the shaded gson class files included in the library are not properly loaded by our module loading code (and possibly in general? the root cause of the bug is unclear at time of writing but it does not appear to be present in all uses of this library). This requires us to use an older version of Nimbus OAuth2 OIDC SDK as well.

Second, the aforementioned shaded gson uses reflection internally, and is used in unpredictable places in these libraries (e.g. constructors). This is extremely unfriendly to our usage of the security manager. In order to make the scope of permission grants as narrow as possible, we shadow nimbus-jose-jwt in order to insert `AccessController.doPrivileged` calls at the appropriate points, given the usage of gson is relatively contained.

This approach was chosen over other approaches given 1) the relative simplicity given the implementation of the library, and 2) the complexity involved in safely using the library any other way - as one example, gson is used frequently in `toString()` methods, which are frequently called implicitly, especially in combination with logging which may mask security manager exceptions from being surfaced in tests. All of the code we intercept should be re-evaluated when this library is next upgraded.

AthenaEryma committed:
Update OAuth2 OIDC SDK

AthenaEryma committed:
WIP

AthenaEryma committed:
Merge branch 'refs/heads/main' into update/oauth2-oidc-sdk

# Conflicts:
#	x-pack/plugin/core/build.gradle

AthenaEryma committed:
Merge branch 'refs/heads/main' into update/oauth2-oidc-sdk

AthenaEryma committed:
remove nimbus require

AthenaEryma committed:
Merge branch 'refs/heads/main' into update/oauth2-oidc-sdk

gwbrown commented on x-pack/plugin/security/src/main/java/module-info.java:
```diff
@@ -49,6 +48,7 @@
     requires oauth2.oidc.sdk;
     requires org.slf4j;
     requires unboundid.ldapsdk;
+    // requires com.nimbusds.jose.jwt;
```
If I leave this commented/omitted, IntelliJ is unhappy as classes from this module are used - makes sense, though Gradle doesn't find a problem(tests still fail on security manager stuff but that's separate). 

If I uncomment this, the Gradle build fails and IntelliJ is differently unhappy, because it can't find the module (despite happily taking me to the definition if I cmd-click it).

AthenaEryma committed:
trying the require in core

AthenaEryma committed:
Merge remote-tracking branch 'origin/main' into update/oauth2-oidc-sdk

AthenaEryma committed:
Merge remote-tracking branch 'origin/main' into update/oauth2-oidc-sdk

AthenaEryma committed:
update to jose-jwt 9.39.1, fixing all the module issues -_-

AthenaEryma committed:
Merge branch 'refs/heads/main' into update/oauth2-oidc-sdk

AthenaEryma committed:
finally got this working, thanks chris

AthenaEryma committed:
Merge remote-tracking branch 'origin/main' into update/oauth2-oidc-sdk

AthenaEryma committed:
9.39.3 patch

AthenaEryma committed:
Revert "9.39.3 patch"

This reverts commit 893495dc6ab149b30ec489edb75bd746bfa8b672.

AthenaEryma committed:
back to 9.37.3

AthenaEryma committed:
Merge branch 'refs/heads/main' into update/oauth2-oidc-sdk

AthenaEryma committed:
cleanup

gwbrown commented:
@rjernst @jakelandis This is the PR in question with the security manager issues, run the failing tests with `./gradlew :x-pack:plugin:security:qa:jwt-realm:javaRestTest`. Failure logs have to be dug out of the build directory so I've uploaded one from a local run of mine [here](https://gist.github.com/gwbrown/7afde7dfa98584a74eafac636a9b2bb0). There's a few different stack traces and I'm not sure which is most important, but the security manager issues are all complaining about `java.lang.RuntimePermission "accessDeclaredMembers"`, which should be granted to the Security module as a whole as well as a codebase for `com.nimbusds.jose.shaded.gson.internal.ConstructorConstructor` specifically (see the security policy and codebases files). That class is present in all of the failure stack traces, as is `JwtAuthenticator.lambda$authenticate$0(JwtAuthenticator.java:75)`, which is an `AccessController.doPrivileged` call.

It's possible that I've misunderstood something - I don't feel like I have a great handle on this stuff, so please if something looks off do call it out.

rjernst commented:
Looking at the stack trace:
```
Caused by: java.security.AccessControlException: access denied ("java.lang.RuntimePermission" "accessDeclaredMembers")
	at java.security.AccessControlContext.checkPermission(AccessControlContext.java:488) ~[?:?]
	at java.security.AccessController.checkPermission(AccessController.java:1085) ~[?:?]
	at java.lang.SecurityManager.checkPermission(SecurityManager.java:411) ~[?:?]
	at java.lang.Class.checkMemberAccess(Class.java:3250) ~[?:?]
	at java.lang.Class.getDeclaredConstructor(Class.java:2951) ~[?:?]
	at com.nimbusds.jose.shaded.gson.internal.ConstructorConstructor.newDefaultConstructor(ConstructorConstructor.java:212) ~[?:?]
	at com.nimbusds.jose.shaded.gson.internal.ConstructorConstructor.get(ConstructorConstructor.java:120) ~[?:?]
	at com.nimbusds.jose.shaded.gson.internal.bind.MapTypeAdapterFactory.create(MapTypeAdapterFactory.java:126) ~[?:?]
	at com.nimbusds.jose.shaded.gson.Gson.getAdapter(Gson.java:556) ~[?:?]
	at com.nimbusds.jose.shaded.gson.Gson.toJson(Gson.java:834) ~[?:?]
	at com.nimbusds.jose.shaded.gson.Gson.toJson(Gson.java:812) ~[?:?]
	at com.nimbusds.jose.shaded.gson.Gson.toJson(Gson.java:759) ~[?:?]
	at com.nimbusds.jose.shaded.gson.Gson.toJson(Gson.java:736) ~[?:?]
	at com.nimbusds.jose.util.JSONObjectUtils.toJSONString(JSONObjectUtils.java:527) ~[?:?]
	at com.nimbusds.jose.Header.toString(Header.java:317) ~[?:?]
	at java.lang.StringConcatHelper.stringOf(StringConcatHelper.java:465) ~[?:?]
	at org.elasticsearch.xpack.security.authc.jwt.JwtRealm.lambda$authenticate$2(JwtRealm.java:269) ~[?:?]
	at org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:62) ~[elasticsearch-8.16.0-SNAPSHOT.jar:?]
	at org.elasticsearch.action.ActionListener$2.onFailure(ActionListener.java:257) ~[elasticsearch-8.16.0-SNAPSHOT.jar:?]
	at org.elasticsearch.xpack.security.authc.jwt.JwtAuthenticator.doAuthenticate(JwtAuthenticator.java:101) ~[?:?]
	at org.elasticsearch.xpack.security.authc.jwt.JwtAuthenticator.lambda$authenticate$0(JwtAuthenticator.java:75) ~[?:?]
	at java.security.AccessController.doPrivileged(AccessController.java:319) ~[?:?]
```

Based on what you described, I suspect the problematic frames are those from the server jar, like this:
```
at org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:62) ~[elasticsearch-8.16.0-SNAPSHOT.jar:?]
```

Perhapse the doPrivileged needs to be moved into this JwtRealm lambda?
```
	at org.elasticsearch.xpack.security.authc.jwt.JwtRealm.lambda$authenticate$2(JwtRealm.java:269) ~[?:?]
```

AthenaEryma committed:
Merge branch 'refs/heads/main' into update/oauth2-oidc-sdk

# Conflicts:
#	x-pack/plugin/core/build.gradle

AthenaEryma committed:
Merge remote-tracking branch 'origin/main' into update/oauth2-oidc-sdk

AthenaEryma committed:
add doPrivileged call

AthenaEryma committed:
Merge remote-tracking branch 'origin/main' into update/oauth2-oidc-sdk

AthenaEryma committed:
one more doPrivileged

AthenaEryma committed:
Merge remote-tracking branch 'origin/main' into update/oauth2-oidc-sdk

AthenaEryma committed:
Add unit test reproducing crypto threadpool queue overflow

AthenaEryma committed:
Fix cache issues on crypto threadpool overflow

AthenaEryma committed:
Revert "Fix cache issues on crypto threadpool overflow"

This reverts commit 0556a5bf37170175bd301651a06338b07df046ae.

AthenaEryma committed:
Revert "Add unit test reproducing crypto threadpool queue overflow"

This reverts commit 6ccacc99e692a994405c1b2147d8e0fb93cb8fef.

AthenaEryma committed:
Merge remote-tracking branch 'origin/main' into update/oauth2-oidc-sdk

AthenaEryma committed:
WIP

AthenaEryma committed:
Merge remote-tracking branch 'origin/main' into update/oauth2-oidc-sdk

jakelandis committed:
passes the testFailureOnExpiredJwt

AthenaEryma committed:
Disable security manager for JWT & OIDC unit tests

We rely on the integ tests to catch stuff anyway

AthenaEryma committed:
Fix failure in unit test

AthenaEryma committed:
Wrap jose access in JWT `ESIntegTestCase`s

AthenaEryma committed:
Merge remote-tracking branch 'origin/main' into update/oauth2-oidc-sdk

AthenaEryma committed:
Expand wrapper to include nimbus

AthenaEryma committed:
Fix typo + wrap one more thing

AthenaEryma committed:
Another wrapper + a bit of cleanup

AthenaEryma committed:
spotless

AthenaEryma committed:
Include system property necessary for unit tests

AthenaEryma committed:
Put testNoSecurityManager next to test

AthenaEryma committed:
General cleanup

AthenaEryma committed:
Revert "Put testNoSecurityManager next to test"

This reverts commit 66fb729a810b237544b885bb330e3a00797708d0.

AthenaEryma committed:
Revert "General cleanup"

This reverts commit 4e96c29a006cfec6aa853ee0a9376c794549bcc1.

AthenaEryma committed:
Remove comment

AthenaEryma committed:
Merge remote-tracking branch 'origin/main' into update/oauth2-oidc-sdk

elasticsearchmachine commented:
Pinging @elastic/es-security (Team:Security)

jakelandis commented on distribution/tools/server-cli/src/main/java/org/elasticsearch/server/cli/ServerProcessBuilder.java:
```diff
@@ -108,6 +108,7 @@ private List<String> getJvmArgs() {
             esHome.resolve("lib").toString(),
             // Special circumstances require some modules (not depended on by the main server module) to be explicitly added:
             "--add-modules=jdk.net", // needed to reflectively set extended socket options
+            "--add-modules=jdk.crypto.ec", // this module is unnecessarily required by com.nimbus.jose.jwt
```
well, this awkward... https://bugs.openjdk.org/browse/JDK-8312267
might want to run the integration tests with Java22 and ensure no scary warnings are emitted.
also, we should run the FIPS tests in CI (IIRC just need to add the label) to ensure no issues when we change the default crypto provider(s).

rjernst replied:
I think this will cause problems later. If the module is unnecessary, can nimbus be convinced to omit it from its module-info, or can we vender the library and adjust the module-info ourselves?

mark-vieira replied:
> or can we vender the library and adjust the module-info ourselves?

This should be an absolute last resort. We have enough forks, special packaging, etc of various libraries.

rjernst replied:
If nimbus does not remove the dependency then I think it's our only option to avoid warnings from the jdk (and then breakage when the module is actually removed).

gwbrown replied:
This turned out not to be necessary; this was for a different version of nimbus-jose-jwt that ended up not being usable for other reasons.

jakelandis commented on x-pack/plugin/security/build.gradle:
```diff
@@ -435,3 +440,33 @@ tasks.named("internalClusterTest").configure {
+// These unit tests are run without the security manager because they use, directly or
+// indirectly, nimbus-jose-jwt code, which uses gson internally, which is not friendly
+// to the security manager. Note that we do not disable the security manager for
+// any integration tests, any failures of which should be taken very seriously.
```
discussed in slack a potential approach to not need to do this.

jakelandis commented on x-pack/plugin/security/lib/jose-wrapper/src/main/java/org/elasticsearch/nimbus/NimbusWrapper.java:
```diff
@@ -0,0 +1,106 @@
+
+/**
+ * This class wraps the operations requiring access in {@link AccessController#doPrivileged(PrivilegedAction)} blocks.
+ * Can't do these operations inline with giving too much access due to how the security manager calculates the stack for lambda expressions.
```
```suggestion
 * Can't do these operations inline without giving too much access due to how the security manager calculates the stack for lambda expressions.
```

jakelandis commented on x-pack/plugin/security/lib/jose-wrapper/src/main/java/org/elasticsearch/nimbus/NimbusWrapper.java:
```diff
@@ -0,0 +1,106 @@
+            if (e.getException() instanceof ParseException ex) {
+                throw ex;
+            } else {
+                throw new RuntimeException(e.getException());
```
We should probably use ElasticsearchException since it can serialize itself, even if we never expect this to happen. (same below).

jakelandis commented on x-pack/plugin/security/lib/jose-wrapper/src/main/java/org/elasticsearch/nimbus/NimbusWrapper.java:
```diff
@@ -0,0 +1,106 @@
+            } else if (exception.getCause() instanceof JOSEException e) {
+                throw e;
+            } else {
+                throw new ElasticsearchException(exception);
```
Can you add a short message to the exception, it can be handy in some cases where don't have direct access to the stack trace.

AthenaEryma committed:
Remove unnecessary no-op module

This require was added in a later version of nimbus-jose-jwt and was left over from the process of figuring out which version was usable.

AthenaEryma committed:
First pass at forbidden APIs

AthenaEryma committed:
Ah, that's how you reference inner classes

AthenaEryma committed:
That's all of them (in nimbus-jose-jwt)

AthenaEryma committed:
Just cut this knot

AthenaEryma committed:
Merge remote-tracking branch 'origin/main' into update/oauth2-oidc-sdk

AthenaEryma committed:
Clean up comments

AthenaEryma committed:
Policy cleanup

AthenaEryma committed:
Bit more cleanup

gwbrown commented:
`part-1-fips` failure is https://github.com/elastic/elasticsearch/issues/111788

@elasticmachine run elasticsearch-ci/part-1-fips

gwbrown commented:
@jakelandis I was working through the thing we discussed with forbidden patterns (which I yes-and'ed into forbidden APIs, for completeness) and saw a way we could shadow nimbus-jose-jwt to inject `doPrivileged` calls in the right places. I understand the concerns raised here about doing this, but the modifications are *very* simple and this is a huge mess otherwise that's quite difficult to contain (look at the [forbidden APIs list](https://github.com/elastic/elasticsearch/blob/cb320bc187029e3ae664096436286f4ba866cf8d/x-pack/plugin/security/forbidden/jwt-signatures.txt) necessary to contain the problem Nikolaj raised, and that's *just* `nimbus-jose-jwt`, not even looking at the OIDC SDK which would also need a similar set).

This lets us do away with modifications to the security manager for tests, the bizarre `NimbusWrapper` class, and is more or less bulletproof (especially compared to the fragility of the other solution, even papered over with forbidden APIs checks).

What do you think of this approach?

jakelandis commented:
>  and saw a way we could shadow nimbus-jose-jwt to inject doPrivileged calls in the right places. I understand the concerns raised here about doing this, but the modifications are very simple 

I don't have a strong opinion to keep what we have, but also unsure how that is possible short of using AOP (Aspect Oriented Programming), which I do have some strong opinions. Can you share some more detail on the alternative approach ?

gwbrown commented on x-pack/plugin/security/lib/nimbus-jose-jwt-fixed/src/main/java/com/nimbusds/jose/util/JSONStringUtils.java:
```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
```
I'm not sure if this is the right thing to do with the copyright headers with these files that are copied almost wholesale from another library.

rjernst replied:
Normally we would retain the original copyright header, see for example FastMath.java

rjernst commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/jwt/JwtSignatureValidatorTests.java:
```diff
@@ -266,7 +266,7 @@ public void testJwtSignVerifyPassedForAllSupportedAlgorithms() {
                 helpTestSignatureAlgorithm(signatureAlgorithm, false);
             } catch (Exception e) {
-                fail("signature validation with algorithm [" + signatureAlgorithm + "] should have succeeded");
+                throw new RuntimeException("signature validation with algorithm [" + signatureAlgorithm + "] should have succeeded", e);
```
nit: this could still be AssertionError

rjernst commented on x-pack/plugin/security/lib/nimbus-jose-jwt-fixed/src/main/java/com/nimbusds/jose/util/JSONObjectUtils.java:
```diff
@@ -0,0 +1,519 @@
+ * 3) This comment and the license comment
+ */
+@SuppressWarnings({ "unchecked", "rawtypes" })
+public class JSONObjectUtils {
```
I understand the desire to avoid lots of doPrivileged calls. However, this seems quite a bit more fragile than other times we have replaced classes from dependencies. For example, with the hdfs client we replaced the shutdown manager with noop methods. In contrast, here many method implementations are copied, and if in future updates they are modified, we might still have the old implementation. It seems like a bigger risk, I lean towards having the doPrivileged calls (though admittedly I did not look at the previous PR iteration, so I don't know how pervasive the doPrivs were).

gwbrown replied:
It's less that the doPriv calls were pervasive and more that a *lot* of things end up calling into one of these methods, including constructors and `toString` methods, to the point where it's not just "pervasive calls", it's "it's legit really hard to add any new code without breaking something".

To take an example @n1v0lg provided (thanks!), consider these two lines:
```
logger.info("Claim set small whoops: {}", jwtAuthenticationToken.getJWTClaimsSet());
logger.info("Claim set big whoops: " + jwtAuthenticationToken.getJWTClaimsSet());
```

Small whoops ends up logging a message like:
`Claim set small whoops: [!!!com.nimbusds.jwt.JWTClaimsSet@15ee0d6d=>java.security.AccessControlException:access denied ("java.lang.RuntimePermission" "accessDeclaredMembers")!!!]`

Big whoops throws an exception. It's very, very, very easy to accidentally hit this, and not even necessarily cause a test failure about it. While we can work around that with forbiddenAPIs (which is what I started doing and can revert back to if we would really, *really* rather do it that way), it's a huge pain in the butt now and any time we have to do anything with the JWT or OIDC realms going forward. On top of that, the forbiddenAPIs list is made by me going through the libraries in question with IntelliJ, tracing all the various ways that `JSONObjectUtils` can be called into. This would need to be updated manually every time we update these dependencies, and less obviously so - at least this is likely to trigger a compilation failure if there are significant changes, whereas a forbiddenAPIs-based solution not being properly updated when the dependency would result in silently permitting us to ship one or both of the "whoops" cases above.

All the options suck. I'm open to being talked into the idea that the forbiddenAPIs-based solution sucks less than vendoring, but it's not good.

jakelandis replied:
I generally like this approach better. However, rather than removing the original class then replacing functionality with wrap(copy/paste), can we instead move the original class to different package then replace the functionality with wrap(delegate(method)) ?

So instead of duplicating the source code, we keep their original source but in a different package name and just wrap the calls to the original code inside the doPrivilged ? It should surface any Java API breaking changes to these methods as compile time issues the next version bump. 

This won't help with the 2 private static method/members, but those seem unlikely to change and we if we really want to (i don't think it is necessary) we could use the elevated privs to delegate to them too via reflections. It might worth a unit test or two to use reflections and mocks to help ensure the same behavior between the original and what is copied, or maybe just a comment in the build.gradle to double check those on a version bump.

gwbrown replied:
Ooh, I like that idea, I think it addresses the worst of the concerns Ryan raised. I'll get that spun together, thanks.

gwbrown replied:
@jakelandis I've put together an approach with `JSONObjectUtils` (and `JSONStringUtils`) copied verbatim (sans formatting) into `InnerJSONObjectUtils` and produced wrappers under the original names. It's not quite as clean as it maybe could be since everything is static, but I think it's still an improvement.

AthenaEryma committed:
Break modifications out into "wrapper" classes

AthenaEryma committed:
License headers check

AthenaEryma committed:
JSONStringUtils javadoc

AthenaEryma committed:
cleanup

AthenaEryma committed:
Merge remote-tracking branch 'origin/main' into update/oauth2-oidc-sdk

rjernst commented on x-pack/plugin/security/lib/nimbus-jose-jwt-fixed/src/main/java/com/nimbusds/jose/util/InnerJSONObjectUtils.java:
```diff
@@ -0,0 +1,515 @@
+ * @version 2022-08-19
+ */
+@SuppressWarnings({ "unchecked", "rawtypes" })
+public class InnerJSONObjectUtils {
```
Rather than copy the source, we should be able to rename the existing classfile when pulling in the dependency?
https://gradleup.com/shadow/configuration/relocation/#filtering-relocation

gwbrown replied:
Doesn't shadowJar rewrite the rest of the JAR to "correctly" reference the moved class when doing that? If I've misunderstood on that point please do correct me, I think that would be better it's just not how I understood that part of shadowJar to work.

rjernst replied:
You are right, it does. However, I think we can still workaround that. We essentially want 3 parts to the final jar:
* The original jar minus these two classes
* These two classes renamed
* The two replacement classes you are adding in

The first two bullets can be separate shadowjar tasks, and ultimately everthing from these 3 parts will go into jar.

gwbrown replied:
Aha, so essentially several intermediate gradle tasks to build a few temporary JARs with the right properies, then merge them into a single final target? Yeah, I think I see how that can work, let me give it a shot.

gwbrown replied:
@rjernst This is done, I agree this is an improvement even though shadowJar gave me a bit of a headache about it. Thanks!

AthenaEryma committed:
Reconfigure the build to reference original class files

AthenaEryma committed:
We have json serializiation at home

AthenaEryma committed:
Merge remote-tracking branch 'origin/main' into update/oauth2-oidc-sdk

gwbrown commented on x-pack/plugin/security/lib/njj-remapped-intermediate/build.gradle:
```diff
@@ -0,0 +1,28 @@
+}
+
+tasks.named('shadowJar').configure {
+  // Attempting to exclude all of the classes we *don't* move here ought to be possible per the
+  // shadowJar docs, but actually attempting to do so results in an empty JAR. So we'll do that as a
+  // separate step (see njj-moved-utils-only)
```
Seriously, I spent some quality time with the shadowJar source trying to figure out why but all the include-based strategies to define which files to include resulted in an empty JAR, and exclude-based ones would require listing out a bunch of files individually since there's a bunch of other stuff in the `com.nimbusds.jose.util.JSONStringUtils` package. I still don't understand why; I think there must be some bug with the relocate implementation and include directives. Breaking it out into two steps like this solves the problem relatively cleanly, but I welcome anyone who can figure out the correct incantation to do the right thing in one step - I think it's possible.

gwbrown replied:
To clarify, `exclude 'com/nimbus/**'` also results in an empty JAR, even though it's exclude-based, regardless of where it's put in the process.

gwbrown commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtUtil.java:
```diff
@@ -11,7 +11,6 @@
 import com.nimbusds.jose.jwk.JWK;
 import com.nimbusds.jose.jwk.JWKSet;
 import com.nimbusds.jose.util.Base64URL;
-import com.nimbusds.jose.util.JSONObjectUtils;
```
IntelliJ wasn't happy about this, I'm not quite sure why - Gradle didn't mind. Easy enough to replace, and none of the other classes seemed to have an issue.

jakelandis commented on x-pack/plugin/security/lib/nimbus-jose-jwt-fixed/build.gradle:
```diff
@@ -0,0 +1,43 @@
+/*
```
nit: can we call this `modified` instead of `fixed` ? Fixed implies we fixed a bug, but really we just modified the original to support the security manager.

jakelandis reviewed (commented):
Looking good just a couple minor comments.

AthenaEryma committed:
Adjust exception handling per review

AthenaEryma committed:
Rename directories + spotless

AthenaEryma committed:
Merge remote-tracking branch 'origin/main' into update/oauth2-oidc-sdk

gwbrown commented:
Thanks for the review @jakelandis, I've adjusted the exception handling and directory names.

jakelandis reviewed (approved):
LGTM

thanks for the iterations!

jakelandis commented:
A quick summary of why this version bump was so involved:

* need to update nimbus and the oidc libraries
* newer versions of nimbus use gson
* gson makes heavy use of reflections
* reflections is not allowed per our security manager policy
* we don't want to generally give permissions for reflections
* to constrain permissions we need wrap calls in doPrivileged
* ran into issues attempting to give explicit permissions to the doPrivilged 
* believe that issues are due to how the security manager calculates the stack for doPrivileged calls from lambdas
* additionally, concerned that future usages will forget about the security manager needs
* decided to rely on wrapping and codebase grants via the security policy
* codebase grants (for us) need to be isolated to jar's
* need to make the doPrivileged call from a jar that is codebase granted  
* need to shadow nimbus to wrap relevant calls to doPrivileged
* need to workaround some odd behavior while shadowing

gwbrown commented:
@mark-vieira Apologies, feel free to disregard that review, misclicked. (Of course, if you have feedback, it's welcome as always.)
